packages/blockchain/examples/optimistic.ts

```diff
@@ -0,0 +1,143 @@
+  )
+  // Blockchain casper Head: 0 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3
+
+  // allows any block > head + 1 to be put as non canonical block
+  await blockchain.putBlock(block3, { canonical: true })
+  headBlock = await blockchain.getCanonicalHeadBlock()
+  blockByHash = await blockchain.getBlock(block3.hash()).catch((e) => null)
+  blockByNumber = await blockchain.getBlock(block3.header.number).catch((e) => null)
+  console.log(
+    `putBlock ${block3.header.number} ${bytesToHex(block3.hash())} byHash: ${blockByHash ? true : false} byNumber: ${blockByNumber ? true : false} headBlock=${headBlock.header.number}`,
+  )
+  // putBlock 15537895 0x4263ec367ce44e4092b79ea240f132250d0d341639afbaf8c0833fbdd6160d0f byHash: true byNumber: true headBlock=0
+
+  let hasBlock1, hasBlock2, hasBlock3
+  hasBlock1 = (await blockchain.getBlock(block1.header.number).catch((e) => null)) ? true : false
+  hasBlock2 = (await blockchain.getBlock(block2.header.number).catch((e) => null)) ? true : false
+  hasBlock3 = (await blockchain.getBlock(block3.header.number).catch((e) => null)) ? true : false
+  console.log(
+    `canonicality: head=${headBlock.header.number}, 0 ... ${block1.header.number}=${hasBlock1} ${block2.header.number}=${hasBlock2} ${block3.header.number}=${hasBlock3} `,
+  )
+  //   canonicality: head=0, 0 ... 15537893=false 15537894=false 15537895=true
+
+  await blockchain.putBlock(block2, { canonical: true })
+  headBlock = await blockchain.getCanonicalHeadBlock()
+  blockByHash = await blockchain.getBlock(block2.hash()).catch((e) => null)
+  blockByNumber = await blockchain.getBlock(block2.header.number).catch((e) => null)
+  console.log(
+    `putBlock ${block2.header.number} ${bytesToHex(block2.hash())} byHash: ${blockByHash ? true : false} byNumber: ${blockByNumber ? true : false} headBlock=${headBlock.header.number}`,
+  )
+  // putBlock 15537894 0x6c33728cd8aa21db683d94418fec1f7ee1cfdaa9b77781762ec832da40ec3a7c byHash: true byNumber: true headBlock=0
+
+  hasBlock1 = (await blockchain.getBlock(block1.header.number).catch((e) => null)) ? true : false
+  hasBlock2 = (await blockchain.getBlock(block2.header.number).catch((e) => null)) ? true : false
+  hasBlock3 = (await blockchain.getBlock(block3.header.number).catch((e) => null)) ? true : false
+  console.log(
+    `canonicality: head=${headBlock.header.number}, 0 ... ${block1.header.number}=${hasBlock1} ${block2.header.number}=${hasBlock2} ${block3.header.number}=${hasBlock3} `,
+  )
+  // canonicality: head=0, 0 ... 15537893=false 15537894=true 15537895=true
+
+  // 1. We can put any post merge block as 4444 anchor by using TTD as parentTD
+  // 2. For pre-merge blocks its prudent to supply correct parentTD so as to respect the
+  //    hardfork configuration as well as to determine the canonicality of the chain on future putBlocks
+  await blockchain.putBlock(block1, { parentTd: chainTTD })
```

Comment:
blockvalidation depends on parent availability in the blockchain and not on any other factor, for optimistic blocks it doesn't happen.

so lets say this was an optimistic block i.e. its parent was not in blockchain, validate blocks will be skipped. if you add canonical=true here, it will add the number => hash index and will run the head update rules (depending on if this is pow block or pos block) since parentTd has been specified and it will assume that it parent chain is confirmed and doesn't need to be added to blockchain

Reply:
> blockvalidation depends on parent availability in the blockchain

This is too implicit behavior. If we have set `validateBlocks = true`, blocks should be validated. Period. There shouldn't be inner auto-triggered exceptions from this rule (except maybe for the very explicit task of setting a checkpoint), otherwise one could not rely on the blockchain functionality anymore.

If one wants to deactivate block validation, we should add an option `blockchain.setBlockValidation(false)` or similar.

Guess this should then be accompanied by a flag for blockchain (if we go this route) to indicate that we might deal with a not-fully validated blockchain.

At least all these things (the "state" of the blockchain) should be transparent and requestable, and not just be "implicitly there".